File: 21561 - Suvam Tamang/Week 7/148_Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
 
        ListNode middle = middleNode(head);
        ListNode nextofmiddle = middle.next;
 
        middle.next = null;
 
        ListNode left = sortList(head);
 
        ListNode right = sortList(nextofmiddle);
 
        ListNode sortedlist = mergeTwoLists(left, right);
        return sortedlist;
    }
    
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {

        ListNode help = new ListNode(0);
        ListNode temp = help; 
        
        while(list1 != null && list2 != null) {
            if(list1.val < list2.val) {
                temp.next = list1;
                list1 = list1.next;
            } else {
                temp.next = list2;
                list2 = list2.next;
            }
            temp = temp.next;
        }
                
        if(list1 != null) {
            temp.next = list1;
        }
        
        if(list2 != null) {
            temp.next = list2;
        }
        
        return help.next;
    }
    
    public ListNode middleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return slow;
    }
    
}

File: 21553 - Sai Nandan D/Week 7/148_Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:        
        if not head or not head.next: return head
        slow, fast = head, head.next        
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next       
        l = head
        r = slow.next
        slow.next = None
        return self.merge(self.sortList(l), self.sortList(r))
    
    
    def merge(self, h1, h2):
        head = tail = ListNode()
        while h1 and h2:
            if h1.val <= h2.val:
                tail.next = h1
                h1 = h1.next                    
            else:
                tail.next = h2
                h2 = h2.next                    
            tail = tail.next      
        tail.next = h1 or h2
        return head.next
